src/gui/core/window_builder.cpp

```diff
@@ -153,53 +155,72 @@ builder_widget_ptr create_builder_widget(const config& cfg)
 		return std::make_shared<implementation::builder_viewport>(viewport);
 	}
 
-/*
- * This is rather odd, when commented out the classes no longer seem to be in
- * the executable, no real idea why, except maybe of an overzealous optimizer
- * while linking. It seems that all these classes aren't explicitly
- * instantiated but only implicitly. Also when looking at the symbols in
- * libwesnoth-game.a the repeating button is there regardless of this #if but
- * in the final binary only if the #if is enabled.
- *
- * If this code is executed, which it will cause an assertion failure.
- *
- * Its likeley that this happens because some build this as a library file
- * which is then used by the wesnoth executable. For msvc a good try to fix
- * this issue is to add __pragma(comment(linker, "/include:" #TYPE)) or
- * similar in the REGISTER_WIDGET3 macro. For gcc and similar this can only
- * be fixed by using --whole-archive flag when linking this library.
- */
-#if 1
-#define TRY(name)                                                                                                      \
-	do {                                                                                                               \
-		if(const config& c = cfg.child(#name)) {                                                                       \
-			builder_widget_ptr p = std::make_shared<implementation::builder_##name>(c);                                \
-			assert(false);                                                                                             \
-		}                                                                                                              \
-	} while(0)
-
-	TRY(stacked_widget);
-	TRY(scrollbar_panel);
-	TRY(horizontal_scrollbar);
-	TRY(repeating_button);
-	TRY(vertical_scrollbar);
-	TRY(label);
-	TRY(image);
-	TRY(toggle_button);
-	TRY(slider);
-	TRY(scroll_label);
-	TRY(matrix);
-	TRY(minimap);
-	TRY(button);
-	TRY(menu_button);
-	TRY(drawing);
-	TRY(password_box);
-	TRY(unit_preview_pane);
-	TRY(size_lock);
-#undef TRY
-#endif
+	for(const auto& item : builder_widget_lookup()) {
+		if(item.first == "window" || item.first == "tooltip") {
+			continue;
+		}
+		if(const config& c = cfg.child(item.first)) {
+			return item.second(c);
+		}
+	}
+
+	// Force all GUI2 Widgets to be linked into Wesnoth
+	//
+	// We use a library. The linker will only include a module from the library if there is
+	// actually a reference to it. While many widgets are referenced elsewhere, it is **NOT**
+	// the case for some. The following block of code never executes, but the optimize must
+	// **NOT** see that! So we test that the cfg reference is a nullptr (references can never
```

Comment:
You're not testing the reference for nullptr, you're invoking the [config-to-bool conversion](https://github.com/wesnoth/wesnoth/blob/master/src/config.hpp#L138).

Although it's a pretty safe bet that the conversion will always return true anyway.

Reply:
Well, I wondered about that. "safe bet" is not good enough. I'll look around for something which is always "false" but which the compiler cannot see as always false.

Meanwhile, I'll correct the comment.

Reply:
Instead of that, I'd prefer putting the fallback code into a function of its own and disallowing the compiler from optimizing it.

```c++
__attribute__((optimize(0)))
void force_linkage(const config& cfg)
{
    FORCE_LINKAGE(addon_list);
    // etc.
}
```

Reply:
Actually, just looked at the bool() operator and it's perfect for my purposes. The only way for `config::operator bool()` to be false (enabling execution of my dead code) should be to violate one of the base rules of C++ and change `this`.

Reply:
only if `__attribute__(optimize(0))` is in C++11 and then only if the standard requires it be honored

Reply:
It's a GCC extension (so we'd have to check the compiler with a preprocessor macro).